asf_heat_pump_suitability/pipeline/reweight_epc/reweight_epc.py

```diff
@@ -1,95 +1,216 @@
-        .then(pl.lit("A caravan or other mobile or temporary structure"))
-        .alias("property_type")
+        dfs.append(pl.DataFrame({feature: missing}))
+
+    # Generate dummy rows with missing categories
+    dummy_rows = pl.concat(dfs, how="horizontal").with_columns(
+        pl.lit(sample["lsoa"].unique()[0], pl.String).alias("lsoa")
+    )
+
+    # Add id to dummy rows and fill nulls in dummy rows
+    dummy_rows = dummy_rows.with_columns(
+        pl.Series(name="UPRN", values=[f"dummy_{_}" for _ in range(0, len(dummy_rows))])
+    )
+    dummy_rows = _fill_nulls_dummy_df(
+        dummy_rows=dummy_rows, lsoa_marginals=lsoa_marginals, sample=sample
     )
 
-    return df
+    return dummy_rows
 
 
-def add_col_nrooms(df: pl.DataFrame) -> pl.DataFrame:
+def _fill_nulls_dummy_df(
+    dummy_rows: pl.DataFrame, sample: pl.DataFrame, lsoa_marginals: Dict[str, dict]
+) -> pl.DataFrame:
     """
-    Add `number_of_rooms` column to EPC dataset with categories corresponding to those from the census.
+    Fill nulls in dummy rows for each feature with another randomly selected missing category if available. If no categories
+    are missing for a feature, fill with the most common category in the EPC subset.
 
     Args:
-        df (pl.DataFrame): EPC dataset
+        dummy_rows (pl.DataFrame): dummy rows containing missing feature categories for a single LSOA in the EPC dataset
+        sample (pl.DataFrame): EPC dataset subset to a single LSOA
+        lsoa_marginals (Dict[str, dict]): dict of dicts containing target proportions for each feature category for a
+        single LSOA
 
     Returns:
-        pl.DataFrame: EPC dataset with `number_of_rooms` column
+        pl.DataFrame: complete dummy rows for a single LSOA in the EPC dataset
     """
-    return df.with_columns(
-        pl.col("NUMBER_HABITABLE_ROOMS")
-        .map_elements(lambda x: 9 if x > 9 else x, return_dtype=pl.Float32)
-        .cast(pl.Int8)
-        .alias("number_of_rooms")
-    )
+    for feature, marginals in lsoa_marginals.items():
+        if dummy_rows[feature].is_null().any():
+            # Identify missing categories for each feature
+            missing = _generate_list_missing_from_sample(
+                feature=feature, feature_marginals=marginals, sample=sample
+            )
+
+            # Fill nulls with random missing category from feature if exists
+            if len(missing) > 0:
+                dummy_rows = dummy_rows.with_columns(
+                    pl.col(feature).fill_null(random.choices(missing, k=1)[0])
+                )
+            # Otherwise, fill nulls with category with max count in sample
+            else:
+                dummy_rows = dummy_rows.with_columns(
+                    pl.col(feature).fill_null(
+                        sample[feature].value_counts().max()[feature][0]
+                    )
+                )
+
+    return dummy_rows
+
+
+def _generate_list_missing_from_sample(
```

Comment:
I guess you could just have one function that returns two things?

```
sample_cats = set(sample[feature].unique())
target_cats = {k for k in feature_marginals.keys() if feature_marginals.get(k) > 0}
missing_from_sample_cats = list(target_cats.difference(sample_cats))
missing_from_target_cats = list(sample_cats.difference(target_cats))

return missing_from_sample_cats, missing_from_target_cats
```

Reply:
Mm I did consider this but if we put our variable names the wrong way round when calling the function, we can get an erroneous run but it will be undetectable as the code won't error out. Instead it would add the wrong dummy rows to sample and drop the wrong feature categories from target. That would be more difficult to detect in the results. I'll keep as is to avoid these errors.